feats: a little bunch of them. see description.

aizensoosuke opened the pull request:
Implements following improvements:

- **Navigable info line** (left click jumps to file).
- Blocks are displayed on the right **day line**. Each day has one (or several, see next items) dedicated lines.
- Configurable **overlaps** (either on new lines, or with a different warning face).
  - Customize with boolean `org-timeline-overlap-in-new-line` and face `org-timeline-overlap`.
- Configurable **clocks** (either not displayed, or on a dedicated line beginning with "$", or in the same line as other events).
  - Customize with booleans `org-timeline-show-clocked` and `org-timeline-dedicated-clocked-line`.
- On generation of the timeline, display **info for the next block** *for today*, i.e. the currently happening one or the next one.
  The next block is also emphasized with a specific face `org-timeline-next-block`.
    - Customize with boolean `org-timeline-emphasize-next-block` that defaults to nil.
- Optionally show **item's title** inside each block. You can only see the first few characters, as it has to be contained in the block.
  If the item has property `TIMELINE_TEXT`, display its value instead of the item's title.
  - Customize with boolean `org-timeline-show-text-in-blocks` that defaults to nil.
- Visually separate blocks that are adjacent using an **"overline"** face on every other adjacent block.
- Ability to **hide elapsed columns** and only keep the last `org-timeline-keep-elapsed` of them.
  This customizable variable defaults to -1 which tells org-timeline to keep all the elapsed columns.
- Ability to see a **24h cycle in timeline**. Setting `org-agenda-span` to 2 and `org-timeline-keep-elapsed` to a positive value merges today and tomorrow's timelines into 1 (only if the agenda currently shows today and tomorrow).
  That way, you see a 24h cycle starting `org-timeline-keep-elapsed` hours ago.
  - This will work better if you set `org-timeline-start-hour` to be 0. It will probably break (i.e. not show blocks that should appear) if you set this to a different value.
  - When doing this, you can easily see today's entire timeline by hitting `j` in the agenda and asking for yesterday to be shown.
  - I'm planning to make this work with `org-timeline-start-hour` when `org-timeline-keep-elapsed` is disabled (negative).
- Adding a character (for now, ❚ i.e. U+275A) at the beginning of each block's title in order to **differentiate consecutive events**.
  - Customize with `org-timeline-insert-before-title` that defaults to "\u275A".
- In a similar manner to clocks, ability to set a **group** for an item. For example, if an event has the property "TIMELINE_GROUP" set to "academics", it will be displayed on a dedicated line starting with "aca". Clocked entries are actually automatically put in the category "  $". 
- Overlap can be customized per-task. You can set the property `TIMELINE_DO_NOT_OVERLAP` for a task.

--- 

Other changes: 
- Changed the customizable variable that sets the hour at which the timeline starts. It is now `org-timeline-beginning-of-day-hour`.
- Following #20, I have also **shifted everything one char to the right**, for better alignment with the hourline. See commit 764e34c. 
- Commit 4bc235a changes the structure and splits `org-timeline--generate-timeline` in a bunch of different functions that I tried my best to make independant.
- Tests have been added to reflect those new features. I also added a list of implemented tests at the top of org-timeline-test.el.

I have taken some inspiration from [deopurkar's fork](https://github.com/deopurkar/org-timeline/commits/master).
---

That's (not) the result (anymore): 
![image](https://user-images.githubusercontent.com/26968781/107290370-b0c76c80-6a66-11eb-9727-0f1038a09a4f.png)

![image](https://user-images.githubusercontent.com/26968781/108087611-e62c1580-7077-11eb-9345-7328748248ab.png)

Commit:
fix: misusing defvar

Commit:
fix: displaying timeline in a way I prefer.

Commit:
fix: offset when moving to task

Commit:
fix: wrong sign

Commit:
fix: misusing defconst

going back to defvar.

aizensoosuke committed:
fix: remove call to clear-info when block is clicked

also remove a useless line 'cursor-sensor...

aizensoosuke committed:
feat: make info navigable to task's file

aizensoosuke committed:
squash! feat: make info navigable to task's file

aizensoosuke committed:
fix: defvar, not defconst.

aizensoosuke committed:
style: use a structure for tasks when building task list

aizensoosuke committed:
feat: one line per day in timeline

overlapping blocks not handled yet.

aizensoosuke committed:
feat: show clocked items in new line

this is customizable with custom variables:
  - `org-timeline-show-clocked`
  - `org-timeline-clocked-in-new-line`

aizensoosuke committed:
feat: handle overlapping blocks

customizable with:
- `org-timeline-overlap` (a face)
- `org-timeline-overlap-in-new-line` (a bool)
there is a bug when clocks and several overlapping blocks
happen for the same day.

aizensoosuke committed:
fix: remove useless vars

aizensoosuke committed:
fix: check overlap for clocks when no dedicated line

aizensoosuke committed:
feat: show info for nearest block

if a block is currently happening, show its info.
else, if there is a next one, show it.
else, if there is a previous one, show the last one.
finally, if no event today, show an empty info line,
  unless the timeline is completely empty.

aizensoosuke committed:
feat: emphasize nearest block with specific face in timeline

customizable with:
- face `org-timeline-nearest-block`
- option `org-timeline-emphasize-nearest-block`
    that defaults to nil.

aizensoosuke committed:
fix: bug when no block today

aizensoosuke committed:
fix: clocks shouldn't be "nearest blocks".

aizensoosuke committed:
fix: broken move-to-task

aizensoosuke committed:
fix: broken info when today not in timeline

aizensoosuke committed:
fix: next event makes much more sense

next commit will also change the feature to only look
for *next* event.

aizensoosuke committed:
style: make code more readable

this also makes it only look for *next* event
(see previous commit)

aizensoosuke committed:
fix: make "next-task" inserted info clickable

also separate `org-timeline--decorate-info`.
also change name of former `org-timeline--clear-info`

aizensoosuke committed:
fix: typo

aizensoosuke committed:
fix: remove outdated comment

aizensoosuke committed:
fix: credits to https://github.com/deopurkar/org-timeline

aizensoosuke committed:
feat: optionally show item text in block

this is customizable with `org-timeline-show-title-in-blocks'
that defaults to nil.

aizensoosuke committed:
fix: error when block is longer than title

Fuco1 commented:
I'm sorry for taking so long, I had some issues going on that needed my attention.  Over the weekend I'll check out the code and add a review.  Thanks for hanging in there!

aizensoosuke commented:
> I'm sorry for taking so long, I had some issues going on that needed my attention. Over the weekend I'll check out the code and add a review. Thanks for hanging in there!

No problem, there's absolutely no rush! My holidays are ending anyway, I'm probably not going to spend much time coding.

aizensoosuke committed:
feat: make events directly following others visible.

adds an overline to every block, except when they are
directly preceded by an element that has an overline.

this way, even when two blocks are right next to each
other, we can see them.

aizensoosuke committed:
refactor: find next-task directly in org-timeline--list-tasks

aizensoosuke committed:
fix: highlight only hovered block on hover.

aizensoosuke committed:
feat: optionally space-out consecutive events

I think this solution is less than ideal. It was
requested in issue #20.

aizensoosuke committed:
fix: wrong end shortened for consecutive blocks.

aizensoosuke committed:
fix: making space-out-consecutive actually work.

aizensoosuke committed:
feat!: shifting everything 1 block

I'm not sure this something you want @Fuco1, so you can just reverse
this commit if you like. It was requested in #20.

aizensoosuke committed:
feat: easier distinction of consecutive blocks

aizensoosuke committed:
feat: ability to hide elapsed time in timeline

customizable with `org-timeline-hide-elapsed`.
see docstring.

aizensoosuke committed:
fix: dotimes was the right macro

aizensoosuke committed:
feat: only hide elapsed times in day view, for today.

also minor bug fixes.

aizensoosuke committed:
feat: true 'change-of-day-hour' configuration.

see the documentation in the code.
using `org-agenda-span` and `org-timeline-keep-elapsed`,
  merges two day into one to see a full 24h cycle starting
  `org-timeline-keep-elapsed` hours ago.

aizensoosuke commented:
This is getting pretty crowded! 
I'm not sure whether I should split this in several pull requests or not anymore.

aizensoosuke commented:
I'd like to clean up what I did a little before you review it, so I'll remove my request review now and add you back after I do that @Fuco1.

aizensoosuke committed:
refactor!(breaking change?): divide to conquer.

new functions:
- `org-timeline--get-cat`
- `org-timeline--get-no-overlap`
- `org-timeline--goto-block-position`
- `org-timeline--make-basic-block`
- `org-timeline--make-and-insert-block`

new customizable variable:
- `org-timeline-insert-before-title`
  By default a vertical bar, this will be inserted before each block's
  title in order to differenciate consecutive blocks.

aizensoosuke committed:
fix: categories bugs.

aizensoosuke committed:
fix: change of day bug

aizensoosuke committed:
fix: empty day line

aizensoosuke committed:
fix: first block ended up in new line.

this needs to be done in a cleaner way...

aizensoosuke committed:
refactor: cleaning up, hopefully better naming.

aizensoosuke committed:
docs: adding documentation for this pull request.

aizensoosuke committed:
fix: overlapping-at-point goes too far

aizensoosuke committed:
refactor: docs and naming

Also added a FIXME item.

aizensoosuke committed:
squash! refactor: docs and naming

aizensoosuke committed:
feat: show past-scheduled items in the timeline

this is needed to make testing easier, but I don't see a reason for this
not to be normal functionality.

aizensoosuke committed:
fix: was broken when no next task

aizensoosuke committed:
feat: manually running tests with this new file

aizensoosuke committed:
fix: new features reflected in test-helper

aizensoosuke committed:
fix: overlaps work better now

aizensoosuke committed:
fix: (partial fix) reflecting changes in the tests

aizensoosuke committed:
fix: last test's agenda string was invalid

aizensoosuke committed:
fix: clock test now works

aizensoosuke committed:
fix: create all the necessary days when several days in timeline

aizensoosuke committed:
tests: fixing existing tests and adding new ones

aizensoosuke committed:
fix: remove useless cursor-sensor-mode

it is unsupported by older versions of emacs and makes travis fail.

aizensoosuke committed:
feat: allow events to be cut if they overflow timeline

aizensoosuke committed:
fix: better face format

aizensoosuke committed:
fix: do not repeat group name in overlap-lines

aizensoosuke committed:
fix: error when no next task

aizensoosuke committed:
fix: allow to have a very "tight" 24h cycle

aizensoosuke committed:
feat: (and fix) tests for all the new features

all implemented tests pass on my machine (doom emacs 27.1)
except for the ones dealing with the 24h cycle.

aizensoosuke committed:
feat(fix, refactor): begin making the cycle compatible with groups

things still need to be improved, see next commit about tests

aizensoosuke committed:
tests: implement tests for most of the features I added

the last 3 tests still don't pass even with the last commit,
next commits should fix that.

aizensoosuke commented:
@Fuco1 

Hi! 

I hope you're doing well. If you have a little time it would be nice if you could take a quick look at #22 and #23.
It's probably less of a quick look, but if you could review `org-timeline--merge-for-24h-cycle` it would be very nice. I'd especially like to know if you think it's a bad way of doing this, and if it should definitely be done in `org-timeline--list-tasks`.

Other than that I think I'm close to marking this PR as ready for review. I just need to fix 3 tests not passing out of 35, and do a review of the whole code myself. I know there will be unoptimized code and all, but this is my first project in elisp so I'm already happy it works well! And hopefully it can help other people.

Thank you !

aizensoosuke committed:
feat: finalized tests and fixed last non-passing-tests

tests pass on emacs 27.1

aizensoosuke committed:
tests(fix): allowing emacs 27.1 to be tested

marking older versions as optional for now, they all fail.

aizensoosuke commented:
I coded and passed all the tests in emacs 27.1.

I'll mark is as ready for review, but I didn't review the whole code myself.

aizensoosuke committed:
fix: overline needs to be in a list

otherwise it overrides the default font-lock-face e.g. org-timeline-block
